daft/execution/physical_plan.py

```diff
@@ -271,6 +272,131 @@ def join(
+            inputs=broadcaster_partitions + [receiver_part.partition()],
+            partial_metadatas=list(broadcaster_partition_metadatas + [receiver_part.partition_metadata()]),
+            resource_request=ResourceRequest(memory_bytes=size_bytes),
+        ).add_instruction(
+            instruction=execution_step.Join(
+                left_on=left_on,
+                right_on=right_on,
+                how=how,
+                is_swapped=is_swapped,
+            )
+        )
+
+    # Materialize the steps from the broadcaster and receiver sources to get partitions.
+    # As the receiver-side materializations complete, emit new steps to join each broadcaster and receiver partition.
+    stage_id = next(stage_id_counter)
+    broadcaster_requests: deque[SingleOutputPartitionTask[PartitionT]] = deque()
+    broadcaster_parts: deque[SingleOutputPartitionTask[PartitionT]] = deque()
+
+    # First, fully materialize the broadcasting side (broadcaster side) of the join.
+    while True:
+        # Moved completed partition tasks in the broadcaster side of the join to the materialized partition set.
+        while broadcaster_requests and broadcaster_requests[0].done():
+            broadcaster_parts.append(broadcaster_requests.popleft())
+
+        # Execute single child step to pull in more broadcaster-side partitions.
+        try:
+            step = next(broadcaster_plan)
+            if isinstance(step, PartitionTaskBuilder):
+                step = step.finalize_partition_task_single_output(stage_id=stage_id)
+                broadcaster_requests.append(step)
+            yield step
+        except StopIteration:
+            if broadcaster_requests:
+                logger.debug(
+                    "broadcast join blocked on completion of broadcasting side of join.\n broadcaster sources: %s",
+                    broadcaster_requests,
+                )
+                yield None
+            else:
+                break
+
+    # Second, broadcast materialized partitions to receiver side of join, as it materializes.
+    receiver_requests: deque[SingleOutputPartitionTask[PartitionT]] = deque()
+    receiver_parts: deque[SingleOutputPartitionTask[PartitionT]] = deque()
+
+    while True:
+        # Moved completed partition tasks in the receiver side of the join to the materialized partition set.
+        while receiver_requests and receiver_requests[0].done():
+            receiver_parts.append(receiver_requests.popleft())
+
+        # Emit join steps for newly materialized partitions.
+        # Broadcast all broadcaster partitions to each new receiver partition that was materialized on this dispatch loop.
+        for receiver_part in receiver_parts:
+            yield _emit_join_step(broadcaster_parts, receiver_part)
+
+        # We always emit a join step for every newly materialized receiver-side partition, so clear the materialized partition queue.
+        receiver_parts.clear()
```

Comment:
We can probably elide this stateful logic if we make `receiver_parts` ephemeral instead:

```python
while True:
    ...
    receiver_parts = []
    ...

    for receiver_part in receiver_parts:
        yield ....

    ...
    # At end of while loop, we're going to create a new `receiver_part`
    
```

for receiver_part in receiver_parts: